chrome/content/zotero/xpcom/data/feed.js

```diff
@@ -267,7 +298,160 @@ Zotero.Feed.prototype._finalizeSave = Zotero.Promise.coroutine(function* (env) {
+		if (this.cleanupAfter) {
+			let expiredItems = yield this.getExpiredFeedItemIDs();
+			Zotero.debug("Cleaning up read feed items...");
+			if (expiredItems.length) {
+				Zotero.debug(expiredItems.join(', '));
+				yield Zotero.FeedItems.forceErase(expiredItems);
+			} else {
+				Zotero.debug("No expired feed items");
+			}
+		}
+	} catch(e) {
+		Zotero.debug("Error clearing expired feed items.");
+		Zotero.debug(e);
+	}
+});
+
+Zotero.Feed.prototype._updateFeed = Zotero.Promise.coroutine(function* () {
+	var toAdd = [];
+	var createNew = true;
+	if (this._updating) {
+		return this._updating;
+	}
+	let deferred = Zotero.Promise.defer();
+	this._updating = deferred.promise;
+	Zotero.Notifier.trigger('statusChanged', 'feed', this.id);
+	this._set('_feedLastCheckError', null);
+	
+	yield this.clearExpiredItems();
+	try {
+		let fr = new Zotero.FeedReader(this.url);
+		yield fr.process();
+		let itemIterator = new fr.ItemIterator();
+		let item, processedGUIDs = [];
+		while (item = yield itemIterator.next().value) {
+			// Append id at the end to prevent same item collisions from different feeds
+			// when to terminate item retrieval.
+			item.guid += ":" + this.id;
+			if (item.guid == this.lastGUID) {
+				Zotero.debug("Feed#update: last seen item reached (" + this.lastGUID + ")");
+				Zotero.debug(item);
+				// Don't create new items (expired and deleted), but update existing ones
+				createNew = false;
+			}
+			
+			if (processedGUIDs.indexOf(item.guid) != -1) {
+				Zotero.debug("Feed item " + item.guid + " already processed from feed.");
+				continue;
+			}
+			processedGUIDs.push(item.guid);
+			
+			Zotero.debug("New feed item retrieved:", 5);
+			Zotero.debug(item, 5);
+			
+			let feedItem = yield Zotero.FeedItems.getAsyncByGUID(item.guid);
+			if (!feedItem && createNew) {
+				feedItem = new Zotero.FeedItem();
+				feedItem.guid = item.guid;
+				feedItem.libraryID = this.id;
+			} else if(! feedItem.isTranslated) {
+				Zotero.debug("Feed item " + item.guid + " already in library.");
+				Zotero.debug("Updating metadata");
+				yield feedItem.loadItemData();
+				yield feedItem.loadCreators();
+			} else {
+				// Either has been translated or beyond lastGUID
+				continue;
+			}
+			
+			// Delete invalid data
+			delete item.guid;
+			feedItem.fromJSON(item);
+			
+			if (!feedItem.hasChanged()) {
+				Zotero.debug("Feed item " + feedItem.guid + " has not changed");
+				continue
+			}
+			feedItem.isRead = false;
+			toAdd.push(feedItem);
+		}
+	}
+	catch (e) {
+		if (e.message) {
+			Zotero.debug("Error processing feed from " + this.url);
+			Zotero.debug(e);
+		}
+		this._set('_feedLastCheckError', e.message || 'Error processing feed');
+	}
+	if (toAdd.length) {
+		yield Zotero.DB.executeTransaction(function* () {
+			// Save in reverse order
+			for (let i=toAdd.length-1; i>=0; i--) {
+				// Saving currently has to happen sequentially so as not to violate the
+				// unique constraints in itemDataValues (FIXME)
```

Comment:
(By the way, I'm just going through the full diff, so I'm sure I'm commenting on some things from the original code.)